### PR TITLE
Add post about WildFly Docker and new provisioning architecture

### PR DESCRIPTION
**Pull request by jmesnil**
Signed-off-by: Jeff Mesnil <jmesnil@redhat.com>

### CODE REVIEW

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 1]
```diff
@@ -0,0 +1,272 @@
+---
+layout: post
+title: 'Use the wildfly-maven-plugin to create Docker image of your application'
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:3]: s/create/create a/g

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+date: 2022-08-05
+tags: cloud container
+description: This article explains how to use the new capabilities of the wildfly-maven-plugin to build container images.
+author: jmesnil
+---
+
+In this article, we will explain how a developer using the Docker image for WildFly can take advantage of the new capabilities of the `wildlfy-maven-plugin` and the new WildFy runtime image to build their container image.
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:10]: s/WildFy/WildFly

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+## WildFly, Docker, S2I and what are the synergies?
+
+WildFly provides a Docker image at https://quay.io/repository/wildfly/wildfly?tab=info[quay.io/wildfly/wildfly] that provides a vanilla WildFly server based on its default standalone configuration.
+
+WildFly also provides Source-to-image (S2I) images to be able to create an application image from the application source using tools targeting OpenShift.
+
+These two types of images, the "vanilla" Docker image and the OpenShift S2I images, have no connection and no synergy.
+A lot of work has been done in WildFly around provisioning. The S2I images benefitted from this effort but the Docker image did not... until now.
+
+While S2I is a good technology to build application image on OpenShift, we realized that we could achieve a more flexible solution that would benefit other container platforms (Kubernetes, Azure) as well as users running WildFly on premise.
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:24]: s/image/images or s/application/an application/g

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+We decided to focus on a Maven-centric approach to provision WildFly so that Maven (and the user) would be in control of the complete runtime (including the user deployment as well as WildFly runtime).
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:26]: s/as well as WildFly/as well as the WildFly/g

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+This Maven-centric approach can also benefit the users of the vanilla Docker image for WildFly and help them move to a consolidated architecture.
+
+## WildFly Docker image
+
+There are different ways to use the https://quay.io/repository/wildfly/wildfly?tab=info[quay.io/wildfly/wildfly] image to create an *application image* (that would contain both WildFly and your deployments) but the simplest one is to add the deployment archive to this base image and let WildFly deploys it when it starts.
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:31]: s/deploys/deploy

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+----
+
+and it replies `MicroProfile Config quickstart deployed successfully. You can find the available operations in the included README file`.
+
+## WildFly Provisioning Architecture
+
+Let's now move to the new architecture for WildFly.
+Jean-Francois Denise extensively described it in the https://www.wildfly.org/news/2022/04/20/WildFly-s2i-v2-Released/[WildFly S2I new architecture is final!] article and focused on the Source-to-Image (S2I) capabilities of the architecture.
+
+In this article, we will see that this architecture also benefits users that are not using S2I and wants to keep control of the creation of their application images.
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:77]: s/users that/users who/g
s/wants to keep/want to keep/g

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+We can see that there is no change from a caller perspective and the application can still be queried with:
+
+[source,bash]
+----
+curl http://localhost:8080/microprofile-config/
+----
+
+## Moving from WildFly Docker image to Runtime image
+
+Let's resume what is needed to move from the vanilla Docker image to the new runtime image for WildFly:
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:151]: s/resume/review

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+Now that we have moved to the new architecture, what are the benefits of it?
+
+## Capability Trimming
+
+WildFly provides capability trimming with https://docs.wildfly.org/26.1/Bootable_Guide.html#wildfly_layers[layers] so that WildFly is provisioned with only the components (mostly Java archives) that are needed to run your application and nothing more. There are two key benefits with capability trimming:
+
+* It reduces the security risk as you are not subject to security attacks if the affected components are not present in application at all.
+* It reduces the size of the server runtime.
+
+In our example, our `microprofile-config` quickstart requires MicroProfile to run. WildFly provides a convenient `microprofile-platform` that provision everything that is needed to run MicroProfile applications. We can trim our runtime with only this layer by updating the `wildfly-maven-plugin`:
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:165]: s/provision/provisions

Comment (bstansberry) [reply _posts/2022-08-05-wildfly-maven-docker.adoc:165]: s/with only/to only/g

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        <layer>microprofile-platform</layer>
+    </layers>
+</configuration>
+----
+
+If we package again the application with `mvn clean package`, we can notice that the size of the `target/server` went from `250M` to `73M` and a lot of jars that were not needed to run the application are no longer present.
+
+## Packaging Scripts
+
+The `wildfly-maven-plugin` also provides the ability to execute JBoss CLI commands when WildFly is provisioned. This allows to substantially modify the standalone configuration to better fit the application requirements. It has no impact on the application image as these scripts are only invoked during provisioning.
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:185]: s/allows to/allows you to/g

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+Adding Undertow Filters for CORS
+...
+----
+
+With the ability to run CLI scripts when WildFly is provisioned, you are in total control of the configuration of WildFly.
+
+## Feature Packs
+
+WildFly uses feature packs as the building blocks to provision the server.
+The most important feature pack is WildFly's own feature pack: `org.wildfly:wildfly-galleon-pack:26.1.1.Final` to contols the installation of WildFly itself.
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:252]: s/controls/control

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+
+We are also providing additional feature packs to provide additional capabilities to WildFly.
+It is out of scope of this article to list all of them but let's discuss two interesting ones:
+
+* The https://github.com/wildfly-extras/wildfly-cloud-galleon-pack[wildfly-cloud-galleon-pack] provides a set of additional features allowing you to configure a WildFly server to work on the cloud. It adapts WildFly to run on orchestration plaftorms in an optimized way. In particular, it automatically route server logs to the console, it provisions the `health` subsystem to monitor the server healthiness, etc.
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:257]: s/route/routes
s/healthiness/health

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+* The https://github.com/wildfly-extras/wildfly-datasources-galleon-pack[wildfly-datasources-galleon-pack] provides JDBC drivers and datasources for various databases. If you include this feature pack, you only need to specify the layer corresponding to the databases you want to use (e.g. `postgresql-datasource`). You then only need to specify a few environment variables (e.g. DB URL and credentials) at runtime to connect to the database when WildFly is running.
+
+## Conclusion
+
+The `wildfly-maven-plugin` is currently at version 4.0.0.Beta2 with a Final release planned for WildFly 27.
+It builds on top of the experience we gained from the Bootable Jar and provide a compelling architecture to control the full runtime (WildFly + the application deployments) from the application's `pom.xml`.
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:263]: s/provide/provides

File: _posts/2022-08-05-wildfly-maven-docker.adoc [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+The full customization of WildFly (using feature packs, packaging scripts, etc.) is controlled by the developer so that the runtime fits the uer's application.
```
Comment (bstansberry) [inline _posts/2022-08-05-wildfly-maven-docker.adoc:264]: s/uer's/user's/g

**Review by bstansberry — CHANGES_REQUESTED** [review]
General content looks good; I've just made some copy-editing notes.

**Comment by jmesnil** [comment]
thanks @bstansberry!

I open this PR as draft as it was not ready for review and needed more editing. Your review confirmed that! :) 
If you are ok with the structure and content, I'll polish it a bit more and we should be good

**Comment by jmesnil** [comment]
@bstansberry I've addressed all your requests and added https://github.com/wildfly/wildfly.org/compare/54c0b8fc25d9a443feb5e82f9b8ec33687c03acd..b3821e39ca55cd872c376702507249e81dd8d430
I think we can rationalise our effort around our images (Docker, S2I, maybe buildpacks in a future) and wants to start a conversation on wildfly-dev about that. I see this article as a way to aggregate feedbacks on this effort.

**Review by bstansberry — APPROVED** [review]

**Pull request merged**